rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,641 @@
+
+#### Concurrent access to the same resources
+
+Is there, for example, some part of the code that is accessing and writing files from the data folder (or anywhere else) 
+and makes the assumption that it is the sole process actually writing to that file?
+
+#### Using instanceUUID as a unique Kibana process identifier
+
+Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
+Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?
+
+#### Things needing to run only once per Kibana instance
+
+Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
+or starting schedulers? 
+
+An example would be Reporting's queueFactory pooling. As we want to only be running a single headless at a time per 
+Kibana instance, only one worker should have pooling enabled.
+
+### 6.2.2 Identified required changes
+
+#### Reporting
+
+We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
+in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 
+to only have the 'main' worker be pooling for reporting tasks.
+
+#### Telemetry
+
+- Server side fetcher
+  
+The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
+We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
+(although race conditions might occur).
+
+- Tasks storing telemetry data
+  
+We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
+tasks will be registered multiple times.
+
+Note that sending the data multiple times doesn’t have any real consequences, apart from the additional number of ES requests, 
+so this should be considered non-blocking and only an improvement.
+
+- Event-based telemetry
+
+Event-based telemetry may be affected as well. Both the existing one in the Security Solutions team and the general 
+one that is in the works. More specifically, the size of the queues will be multiplied per worker, also growing in the
+amount of network bandwidth used, and potentially affecting our customers.
+
+We could address that by making sure that the queues are held only in the main worker.
+
+#### Task Manager
+
+Currently, task manager does "claims" for jobs to run based on the server uuid. We think this could still work with 
+a multi-process setup - each task manager in the worker would be doing "claims" for the same server uuid, which
+seems functionally the same as setting max_workers to `current max_workers * number of workers`.
+
+However, as a first step we can simply run Task Manager on the main worker. This doesn't completely solve potential
```

Comment:
I would like to add an option to the table. The task manager doesn't need to be using the server uuid for claiming jobs, it just needs a unique identifier. If we can compose something like server_uuid + worker id, this would suffice as well because we can still trace tasks to servers while benefiting clustering if we wish to have each worker run task manager.

cc @gmmorris to keep me honest.

Reply:
Yeah, 100% correct.
We use UUID to make it easier to map an owner ID on a  task back to a Kibana, but we can also generate a TM specific UUIS.